op-chain-ops/interopgen/deployers/implementations.go

```diff
@@ -0,0 +1,101 @@
+)
+
+type DeployImplementationsInput struct {
+	WithdrawalDelaySeconds          *big.Int
+	MinProposalSizeBytes            *big.Int
+	ChallengePeriodSeconds          *big.Int
+	ProofMaturityDelaySeconds       *big.Int
+	DisputeGameFinalityDelaySeconds *big.Int
+	// Release version to set OPSM implementations for, of the format `op-contracts/vX.Y.Z`.
+	Release               string
+	SuperchainConfigProxy common.Address
+	ProtocolVersionsProxy common.Address
+	UseInterop            bool // if true, deploy Interop implementations
+}
+
+func (input *DeployImplementationsInput) InputSet() bool {
+	return true
+}
+
+type DeployImplementationsOutput struct {
+	Opsm                             common.Address
+	DelayedWETHImpl                  common.Address
+	OptimismPortalImpl               common.Address
+	PreimageOracleSingleton          common.Address
+	MipsSingleton                    common.Address
+	SystemConfigImpl                 common.Address
+	L1CrossDomainMessengerImpl       common.Address
+	L1ERC721BridgeImpl               common.Address
+	L1StandardBridgeImpl             common.Address
+	OptimismMintableERC20FactoryImpl common.Address
+	DisputeGameFactoryImpl           common.Address
+}
+
+func (output *DeployImplementationsOutput) CheckOutput() error {
+	return nil
+}
+
+type DeployImplementationsScript struct {
+	Run func(input, output common.Address) error
+}
+
+func DeployImplementations(l1Host *script.Host, input *DeployImplementationsInput) (*DeployImplementationsOutput, error) {
+	output := &DeployImplementationsOutput{}
+	inputAddr := l1Host.NewScriptAddress()
+	outputAddr := l1Host.NewScriptAddress()
+
+	cleanupInput, err := script.WithPrecompileAtAddress[*DeployImplementationsInput](l1Host, inputAddr, input)
+	if err != nil {
+		return nil, fmt.Errorf("failed to insert DeployImplementationsInput precompile: %w", err)
+	}
+	defer cleanupInput()
+
+	cleanupOutput, err := script.WithPrecompileAtAddress[*DeployImplementationsOutput](l1Host, outputAddr, output,
+		script.WithFieldSetter[*DeployImplementationsOutput])
+	if err != nil {
+		return nil, fmt.Errorf("failed to insert DeployImplementationsOutput precompile: %w", err)
+	}
+	defer cleanupOutput()
+
+	implContract := "DeployImplementations"
+	if input.UseInterop {
+		implContract = "DeployImplementationsInterop"
+	}
+	deployScript, cleanupDeploy, err := script.WithScript[DeployImplementationsScript](l1Host, "DeployImplementations.s.sol", implContract)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load %s script: %w", implContract, err)
+	}
+	defer cleanupDeploy()
+
+	opsmContract := "OPStackManager"
+	if input.UseInterop {
+		opsmContract = "OPStackManagerInterop"
+	}
+	if err := l1Host.RememberOnLabel("OPStackManager", opsmContract+".sol", opsmContract); err != nil {
+		return nil, fmt.Errorf("failed to link OPStackManager label: %w", err)
+	}
+
+	// So we can see in detail where the SystemConfig interop initializer fails
```

Comment:
Is there a generic way to register all of the contracts like this? Very likely we will see failures like this in the future as there is more development of the smart contracts

Reply:
I tried to use the `vm.label` to automatically hook up contracts, but the labels are quite inconsistent, and the label emitted by the script is the same, regardless if it's the interop or non-interop variant of a contract, so it's still difficult to auto-attach the right source to an address. More consistency there would help.